### PR TITLE
Fixes to address unintended props reshuffling behaviour

### PR DESCRIPTION
**Pull request by rpkyle**
Proposed changes introduced by this PR:

- The ordering of `props` in Dash components is maintained by providing `object_pairs_hook=OrderedDict` to `json.loads` .
- Handling of `prop_keys` exclusions has been refactored to account for removal of events, and to properly omit wildcards.
- R components now require `children` to be passed as a list or function, as in Dash for Python, and ellipsis is used for wildcards only, as the final formal argument for component functions.
- Component library dependencies are now copied to `inst/deps` rather than `inst/lib` to address an issue with `install_github`.
- R documentation now has additional whitespace to improve readability.
- R (overall) package help files are now available.
- R component handling of `defaultValue` fields is now similar to Dash for Python, fixing issues with rendering of components due to passing incorrect defaults.
- Removed dependency of JavaScript metadata generation for DashR on `version.py`.
- Use modified version of Mark Amery's `byteify` function to avoid storing strings from JSON as Unicode in Python 2.x, so that lists of strings in components should be identical in appearance across versions.
- A new function `assert_valid_wildcards` is called when wildcards are available for a given component, to validate any wildcard input passed in via `...` in the corresponding function formals in R.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -20,3 +20,4 @@ dist
 npm-debug*
 /.tox
 .idea
+package-lock.json
```
Comment (alexcjohnson) [inline .gitignore]: It shouldn't be necessary to ignore this, though there are occasional issues.

Comment (rpkyle) [reply .gitignore]: Sorry about that, will edit to remove!

File: dash/development/_r_components_generation.py [hunk 2]
```diff
@@ -180,15 +159,13 @@ def generate_class_string(name, props, project_shortname, prefix):
...
     # Filter props to remove those we don't want to expose
-    for p in prop_keys:
-        if p.endswith("-*") \
-                or p in r_keywords \
-                or p in ['setProps', 'dashEvents', 'fireEvent']:
-            prop_keys.remove(p)
+    for item in prop_keys[:]:
+        if item.endswith('-*') \
+                or item in r_keywords \
+                or item == 'setProps':
```
Comment (T4rk1n) [inline dash/development/_r_components_generation.py]: could be just:

```python
or item in r_keywords + ['setProps']
```

File: dash/development/_r_components_generation.py [hunk 3]
```diff
@@ -163,7 +142,7 @@ def generate_class_string(name, props, project_shortname, prefix):
         '\'{}\''.format(p)
         for p in prop_keys
         if '*' not in p and
-        p not in ['setProps', 'dashEvents', 'fireEvent']
+        p != 'setProps' + ['**kwargs']
```
Comment (T4rk1n) [inline dash/development/_r_components_generation.py]: That doesn't work.
```python
>>> 'setProps' + ['**kwargs']
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: must be str, not list
```

```suggestion
        p not in ['setProps', '**kwargs']
```

Comment (alexcjohnson) [reply dash/development/_r_components_generation.py]: Oh that's not quite it either... the `p not in [...]` or `p != ...` clause is a filter in an iterator - in the python generator code: 
https://github.com/plotly/dash/blob/e72f483c85aaa91e4316a548ed83ab40a6819452/dash/development/_py_components_generation.py#L118-L126

note the extra `[...]` from immediately after `.join(` to right before `+ ['**kwargs']` that turns this into a list that we can then append `'**kwargs'` to by adding a second list. I'm assuming in R `**kwargs` doesn't mean anything, but there's some other feature to play the role of allowing arbitrary extra named arguments?

**Comment by alexcjohnson** [comment]
> I don't understand the `package-lock.json`, there's no `package.json` in dash. 🔪

Hah too funny! Right, `dash` is the one repo that doesn't use `npm` at all. Oh well, sounds like @rpkyle sorted this out :)

File: dash/development/_r_components_generation.py (hunk 3 above)
Comment (rpkyle) [reply dash/development/_r_components_generation.py]: > I'm assuming in R **kwargs doesn't mean anything, but there's some other feature to play the role of allowing arbitrary extra named arguments?

@alexcjohnson Correct. For the wildcards, I've used ellipsis (aka `\dots`, `...`) for this purpose. It seems to be working to pass in an arbitrary number (including zero) of wildcard arguments.

**Comment by rpkyle** [comment]
> > I don't understand the `package-lock.json`, there's no `package.json` in dash. hocho
> 
> Hah too funny! Right, `dash` is the one repo that doesn't use `npm` at all. Oh well, sounds like @rpkyle sorted this out :)

I'm glad I've somehow fallen into the component management pipeline on the Python side as well, because for all the bumps and bruises, I'm certainly learning a few things.

File: dash/development/_r_components_generation.py (hunk 2 above)
Comment (rpkyle) [reply dash/development/_r_components_generation.py]: This didn't work:

```ts_generation.py", line 164, in generate_class_string
    or item in r_keywords + ['setProps']:
TypeError: unsupported operand type(s) for +: 'set' and 'list'```

**Comment by alexcjohnson** [comment]
> I attempted to address this by wrapping `props.keys()` with `list`;

Yeah, `.keys()` and various other things return iterators instead of lists in py3, that one continually trips me up / annoys me... you've got the correct solution.

**Comment by T4rk1n** [comment]
> File "dash/dash/development/_r_components_generation.py", line 282, in write_help_file for item in prop_keys[:]: TypeError: 'odict_keys' object is not subscriptable

Yes, `dict.keys()` returns a list in python 2 and a generator in python 3 so you need to cast to list.

File: dash/development/component_generator.py [hunk 4]
```diff
@@ -82,13 +89,17 @@ def generate_components(components_source, project_shortname,
     )
 
     with open(os.path.join(project_shortname, 'metadata.json'), 'w') as f:
-        json.dump(metadata, f)
+        json.dump(metadata, f, sort_keys=True)
```
Comment (T4rk1n) [inline dash/development/component_generator.py]: :rewind: Sorting the key reorder the keys, the ordered dict will keep the right order alone.

**Comment by rpkyle** [comment]
@T4rk1n If PRs #439 and #91 look ready to go, mind quickly checking this one also? It should be somewhat straightforward.

**Pull request merged**